chainer/training/trainer.py

```diff
@@ -0,0 +1,285 @@
+        update unless the corresponding trigger object decides to skip the
+        iteration. The order of execution is determined by priorities:
+        extensions with higher priorities are called earlier in each iteration.
+        Extensions with the same priority are invoked in the order of
+        registrations.
+
+        If two or more extensions with the same name are registered, suffixes
+        are added to the names of the second to last extensions. The suffix is
+        ``_N`` where N is the ordinal of the extensions.
+
+        See :class:`Extension` for the interface of extensions.
+
+        Args:
+            extension: Extension to register.
+            trigger (tuple or Trigger): Trigger object that determines when to
+                invoke the extension. If it is None, ``extension.trigger`` is
+                used instead. If the trigger is not callable, it is passed to
+                :class:`IntervalTrigger` to build an interval trigger.
+            priority (int): Invocation priority of the extension. Extensions
+                are invoked in the descending order of priorities in each
+                iteration. If this is None, ``extension.priority`` is used
+                instead.
+            invoke_before_training (bool): If True, the extension is also
+                invoked just before entering the training loop. If this
+                None, ``extension.invoke_before_training`` is used instead.
+                This option is mainly used for extensions that alter the
+                training configuration (e.g., learning rates); in such a case,
+                resuming from snapshots require the call of extension to
+                recover the configuration before any updates.
+
+        """
+        if name is None:
+            name = getattr(extension, 'name', None)
+            if name is None:
+                name = getattr(extension, 'default_name', None)
+                if name is None:
+                    raise TypeError('name is not given for the extension')
+                elif name == 'training':
+                    raise ValueError('the name "training" is reserved')
+
+        if trigger is None:
+            trigger = getattr(extension, 'trigger', None)
+        trigger = trigger_module.get_trigger(trigger)
+
+        if priority is None:
+            priority = getattr(
+                extension, 'priority', extension_module.PRIORITY_READER)
+
+        if invoke_before_training is None:
+            invoke_before_training = getattr(
+                extension, 'invoke_before_training', False)
+
+        modified_name = name
+        ordinal = 0
+        while modified_name in self._extensions:
+            ordinal += 1
+            modified_name = '%s_%d' % (name, ordinal)
+
+        extension.name = modified_name
+        self._extensions[modified_name] = _ExtensionEntry(
+            extension, priority, trigger, invoke_before_training)
```

Comment:
There would be a case of registering one extension instance multiple times to some trainer. Users can even register one extension instance multiple times to one trainer by different names.